### PR TITLE
[Feature] add backdoor attack

### PR DESCRIPTION
**Pull request by xieyxclack**
As the title says.  This thread is inherited from #250, since the checking of Contributor License Agreement could not be passed for some reasons in #250.

Please note that the contributor and author of this pr is @Alan-Qin, and I only help on committing. 
Thanks a lot for the contribution :)

### CODE REVIEW

File: federatedscope/contrib/model/resnet.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
+from federatedscope.register import register_model
```
Comment (DavdGao) [inline federatedscope/contrib/model/resnet.py:1]: If the ResNet is the same with the one in `torchvision.models`, maybe we only need to modify `federatedscope/core/auxiliaries/model_builder.py` rather than copying it here.

File: federatedscope/core/configs/cfg_attack.py [hunk 2]
```diff
@@ -13,6 +13,30 @@ def extend_attack_cfg(cfg):
...
     cfg.attack.attacker_id = -1
 
+    # for backdoor attack
+
+    cfg.attack.edge_path = ''
+    cfg.attack.trigger_path = ''
+    cfg.attack.setting = 'fix'
+    cfg.attack.freq = 10
+    cfg.attack.insert_round = 100000
+    cfg.attack.mean = [0.1307]
```
Comment (DavdGao) [inline federatedscope/core/configs/cfg_attack.py:23]: Why does the mean and std default to 0.1307 and 0.3081? Maybe they should be set according to cfg.data.type？

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+def load_poisoned_dataset_edgeset(data, ctx, mode):
+
+    transforms_funcs = get_transform(ctx, 'torchvision')['transform']
+    load_path = ctx.attack.edge_path
+    if "femnist" in ctx.data.type:
+        if mode == 'train':
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:24]: please import `federatedscope/core/auxiliaries/enums.py` and use MODE.TRAIN  instead

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+                poisoned_edgeset = torch.load(saved_data_file)
+            num_dps_poisoned_dataset = len(poisoned_edgeset)
+
+            for ii in range(num_dps_poisoned_dataset):
+                sample, label = poisoned_edgeset[ii]
+                # (channel, height, width) = sample.shape #(c,h,w)
+                sample = sample.numpy().transpose(1, 2, 0)
+                data['train'].dataset.append((transforms_funcs(sample), label))
+
+        if mode == 'test' or 'val':
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:37]: The same as above

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+                samped_poisoned_data_indices, :, :, :]
+
+            for ii in range(num_poisoned_dataset):
+                sample = saved_southwest_dataset_train[ii]
+                data['train'].dataset.append((transforms_funcs(sample), label))
+
+            logger.info('adding {:d} edge-cased samples in CIFAR-10'.format(
+                num_poisoned))
+
+        if mode == 'test' or 'val':
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:79]: The same as above

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+                poison_testset,
+                batch_size=ctx.data.batch_size,
+                shuffle=False,
+                num_workers=ctx.data.num_workers)
+
+    elif "CIFAR10" in ctx.data.type:
+        target_label = int(ctx.attack.target_label_ind)
+        label = target_label
+        num_poisoned = ctx.attack.edge_num
+        if mode == 'train':
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:59]: The same as above

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        perm = np.random.permutation(
+            len(dataset))[0:int(len(dataset) * inject_portion)]
+
+    dataset_ = list()
+    for i in range(len(dataset)):
+        data = dataset[i]
+
+        if label_type == 'dirty':
+            # all2one attack
+            if mode == 'train':
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:139]: MODE.TRAIN

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+
+                elif 'wanet' in trigger_type and i in perm_cross:
+                    img = selectTrigger(img, width, height, distance, trig_w,
+                                        trig_h, 'wanetTriggerCross', load_path)
+                    dataset_.append((img, data[1]))
+
+                else:
+                    dataset_.append((img, data[1]))
+
+            if mode == 'test' or 'val':
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:159]: MODE.TEST or MODE.VAL

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        inject_portion_train = ctx.attack.poison_ratio
+    else:
+        raise RuntimeError(
+            'Now, we only support the FEMNIST and CIFAR-10 datasets')
+
+    inject_portion_test = 1.0
+
+    load_path = ctx.attack.trigger_path
+
+    if mode == 'train':
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:197]: MODE.TRAIN

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        for iii in range(num_dps_poisoned_dataset):
+            sample, label = poisoned_dataset[iii]
+            poisoned_dataset[iii] = (transforms_funcs(sample), label)
+
+        data['train'] = DataLoader(poisoned_dataset,
+                                   batch_size=ctx.data.batch_size,
+                                   shuffle=True,
+                                   num_workers=ctx.data.num_workers)
+
+    if mode == 'test' or 'val':
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:218]: MODE.TEST or MODE.VAL

File: federatedscope/attack/auxiliary/create_edgeset.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+    ardis_test_dataset = []
+
+    for ii in range(len(images_seven)):
+        ardis_test_dataset.append((images_seven[ii], poisoned_labels[ii]))
+
+    return ardis_test_dataset
+
+
+if __name__ == '__main__':
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/create_edgeset.py:124]: Please remove the main function.

File: federatedscope/attack/auxiliary/poisoning_data.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+import logging
+import os
+
+logger = logging.getLogger(__name__)
+
+
+def load_poisoned_dataset_edgeset(data, ctx, mode):
+
+    transforms_funcs = get_transform(ctx, 'torchvision')['transform']
+    load_path = ctx.attack.edge_path
```
Comment (DavdGao) [inline federatedscope/attack/auxiliary/poisoning_data.py:22]: Please use `data.root` instead, and you can create a sub directory under `data.root`.

**Review by DavdGao — COMMENTED** [review]
Please  see the inline comments

File: federatedscope/attack/auxiliary/create_edgeset.py (hunk 11 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/create_edgeset.py:128]: fix it

File: federatedscope/attack/auxiliary/poisoning_data.py (hunk 12 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/poisoning_data.py:24]: fix it

File: federatedscope/attack/auxiliary/poisoning_data.py (hunk 9 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/poisoning_data.py]: fit it

File: federatedscope/attack/auxiliary/poisoning_data.py (hunk 8 above)
Comment (Alan-Qin) [reply federatedscope/attack/auxiliary/poisoning_data.py]: fix it

File: federatedscope/attack/auxiliary/create_edgeset.py (hunk 11 above)
Comment (Osier-Yi) [reply federatedscope/attack/auxiliary/create_edgeset.py:128]: The main function still here?

> fix it

Comment (Osier-Yi) [reply federatedscope/attack/auxiliary/create_edgeset.py]: fix it!

**Review by Osier-Yi — COMMENTED** [review]
LGTM

**Comment by DavdGao** [comment]
LGTM, @Alan-Qin please remember to finish the following modifications in the future. 
- Add unit test for each backdoor attack algorithm
- Use `data.root` rather than `attack.edge_path`, or set `attack.edge_path='test_data/'` in the corresponding unit test to avoid saving data in the Jenkins server. 
- Download the poison dataset from url rather than manually downloading.
- Update the tutorial

**Review by DavdGao — APPROVED** [review]
